File: src/core/process.ts
import { readFileSync, writeFileSync } from 'node:fs';
import type { FileResult, KeepRule } from '../types.js';
import { mayContainComments, scanSource } from './scan.js';
import { stripComments } from './strip.js';

export interface ProcessOptions {
  keep: readonly KeepRule[];
  collapseBlankLines: boolean;
  write: boolean;
}

export function processFile(file: string, options: ProcessOptions): FileResult {
  const clean: FileResult = { file, removable: [], keptCount: 0, changed: false };

  let source: string;
  try {
    source = readFileSync(file, 'utf8');
  } catch (error) {
    return { ...clean, error: error instanceof Error ? error.message : String(error) };
  }

  if (!mayContainComments(source)) return clean;

  let result;
  try {
    result = scanSource(source, { fileName: file, keep: options.keep });
  } catch (error) {
    return { ...clean, error: error instanceof Error ? error.message : String(error) };
  }

  if (result.ignoredFile) return clean;

  const output = stripComments(source, result.removable, {
    collapseBlankLines: options.collapseBlankLines,
  });
  const changed = output !== source;

  if (changed && options.write) {
    try {
      writeFileSync(file, output, 'utf8');
    } catch (error) {
      return {
        ...clean,
        removable: result.removable,
        keptCount: result.kept.length,
        error: error instanceof Error ? error.message : String(error),
      };
    }
  }

  return {
    file,
    removable: result.removable,
    keptCount: result.kept.length,
    changed,
  };
}

File: packages/python/src/commentless/process.py
from __future__ import annotations

from dataclasses import dataclass
from pathlib import Path

from .scan import may_contain_comments, scan_source
from .strip import strip_comments
from .types import FileResult, KeepRule, ScanOptions


@dataclass(frozen=True, slots=True)
class ProcessOptions:
    keep: tuple[KeepRule, ...] = ()
    collapse_blank_lines: bool = False
    write: bool = False
    docstrings: bool = False


def process_file(file: str, options: ProcessOptions) -> FileResult:
    try:
        source = Path(file).read_text(encoding="utf-8")
    except (OSError, UnicodeDecodeError) as error:
        return FileResult(file=file, error=str(error))

    if not may_contain_comments(source, docstrings=options.docstrings):
        return FileResult(file=file)

    try:
        result = scan_source(
            source,
            ScanOptions(file_name=file, keep=options.keep, docstrings=options.docstrings),
        )
    except Exception as error:
        return FileResult(file=file, error=str(error))

    if result.ignored_file:
        return FileResult(file=file)

    output = strip_comments(
        source,
        result.removable,
        collapse_blank_lines_option=options.collapse_blank_lines,
    )
    changed = output != source

    if changed and options.write:
        try:
            Path(file).write_text(output, encoding="utf-8")
        except OSError as error:
            return FileResult(
                file=file,
                removable=result.removable,
                kept_count=len(result.kept),
                error=str(error),
            )

    return FileResult(
        file=file,
        removable=result.removable,
        kept_count=len(result.kept),
        changed=changed,
    )
